parser-typechecker/src/Unison/Lexer.hs

```diff
@@ -60,74 +97,440 @@ type IsVirtual = Bool -- is it a virtual semi or an actual semi?
 
 makePrisms ''Lexeme
 
+space :: P ()
+space = LP.space CP.space1 (fold <|> LP.skipLineComment "--")
+                           (LP.skipBlockCommentNested "{-" "-}")
+  where
+  fold = P.try $ lit "---" *> P.takeRest *> pure ()
+
+lit :: String -> P String
+lit = P.try . LP.symbol (pure ())
+
+token :: P Lexeme -> P [Token Lexeme]
+token = token' (\a start end -> [Token a start end])
+
+pos :: P Pos
+pos = do
+  p <- P.getPosition
+  pure $ Pos (P.unPos (P.sourceLine p)) (P.unPos (P.sourceColumn p))
+
+-- Token parser: strips trailing whitespace and comments after a
+-- successful parse, and also takes care of emitting layout tokens
+-- (such as virtual semicolons and closing tokens).
+token' :: (a -> Pos -> Pos -> [Token Lexeme]) -> P a -> P [Token Lexeme]
+token' tok p = LP.lexeme space (token'' tok p)
+
+-- Committed failure
+err :: Pos -> Err -> P x
+err start t = do
+  stop <- pos
+  _ <- void CP.anyChar <|> P.eof
```

Comment:
Do you recall what this line is for? Looks like it's just for adjusting the underlying parser's state, but I'm not totally sure what that would be for if we're reporting a failure. Maybe it's for Megaparsec's multi-error support?

Reply:
That line consumes a token and therefore produces a committed failure. So you can have things like `try p <|> err s t` and it will not backtrack to trying something else. I'll add a comment to this effect.

Reply:
okay, added a comment on this